### PR TITLE
Add ``send_chunk`` utility for repeated trace-then-send pattern

### PR DESCRIPTION
**Pull request by rjzamora**
## Description
- Adds simple `send_chunk` utiity
    - A repeated pattern across the codebase.
- Updates `ActorTracer.add_chunk` to actually accept a `TableChunk` instead of a `plc.Table`
    - Avoids un-spilling to record the row-count.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf_polars/tests/streaming/test_tracing.py [hunk 1]
```diff
@@ -4,12 +4,59 @@
...
+from cudf_polars.streaming.actor_graph.tracing import ActorTracer, send_chunk
+
+
+def _table_chunk(spmd_engine) -> TableChunk:
+    context = spmd_engine.context
+    stream = context.get_stream_from_pool()
+    df = DataFrame.from_polars(pl.DataFrame({"x": [1, 2, 3]}), stream)
+    return TableChunk.from_pylibcudf_table(
+        df.table, stream, exclusive_view=True, br=context.br()
+    )
```
Comment (wence-) [inline python/cudf_polars/tests/streaming/test_tracing.py]: No, make this a fixture.


```suggestion
@pytest.fixture
def chunk(spmd_engine: SPMDEngine) -> TableChunk:
    context = spmd_engine.context
    stream = context.get_stream_from_pool()
    df = DataFrame.from_polars(pl.DataFrame({"x": [1, 2, 3]}), stream)
    return TableChunk.from_pylibcudf_table(
        df.table, stream, exclusive_view=True, br=context.br()
    )
```

File: python/cudf_polars/tests/streaming/test_tracing.py [hunk 2]
```diff
@@ -4,12 +4,59 @@
...
+
+
+@pytest.mark.spmd
+def test_actor_tracer_counts_table_chunk_without_table_view(spmd_engine):
+    tracer = ActorTracer()
+    chunk = _table_chunk(spmd_engine)
```
Comment (wence-) [inline python/cudf_polars/tests/streaming/test_tracing.py]:
```suggestion
def test_actor_tracer_counts_table_chunk_without_table_view(chunk):
    tracer = ActorTracer()
```

et seq.

File: python/cudf_polars/tests/streaming/test_tracing.py [hunk 3]
```diff
@@ -4,12 +4,59 @@
...
+@pytest.mark.spmd
+def test_send_chunk_traces_and_sends_message(spmd_engine):
+    context = spmd_engine.context
+    ch_out = context.create_channel()
+    tracer = ActorTracer()
+    chunk = _table_chunk(spmd_engine)
+
+    async def send_and_recv():
+        await send_chunk(context, ch_out, chunk, 11, tracer=tracer)
+        return await ch_out.recv(context)
```
Comment (wence-) [inline python/cudf_polars/tests/streaming/test_tracing.py]: No, this relies on implementation details of the channel. If you want to run send_chunk and `ch_out.recv` concurrently you must do gather them in a `TaskGroup`.

**Review by wence- — DISMISSED** [review]

File: python/cudf_polars/tests/streaming/test_tracing.py [hunk 4]
```diff
@@ -4,12 +4,59 @@
...
+
+    msg = asyncio.run(send_and_recv())
+
+    assert msg is not None
+    assert msg.sequence_number == 11
+    assert TableChunk.from_message(msg, br=context.br()).shape[0] == 3
+    assert tracer.chunk_count == 1
+    assert tracer.row_count == 3
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/cudf_polars/tests/streaming/test_tracing.py:65]: I don’t think all-null and mixed-type chunks are distinct behavioral cases for `send_chunk`. The helper only inspects `TableChunk.shape[0]` and otherwise forwards the chunk unchanged. Those dtype/null edge cases are covered more meaningfully by existing streaming integration tests and TableChunk/channel tests.

**Comment by rjzamora** [comment]
Thanks for the review @wence- Hopefully this is good now.

File: python/cudf_polars/cudf_polars/streaming/actor_graph/repartition.py [hunk 5]
```diff
@@ -220,19 +218,14 @@ async def concatenate_node(
...
+                        df.table,
+                        df.stream,
+                        exclusive_view=True,
+                        br=context.br(),
+                    )
+                    await send_chunk(
+                        context, ch_out, output_chunk, seq_num, tracer=tracer
                     )
                     seq_num += 1
                     del df
```
Comment (TomAugspurger) [inline python/cudf_polars/cudf_polars/streaming/actor_graph/repartition.py:231]: Do you know whether we also need to `del output_chunk`? I don't really recall why we have this `del df`, but presumably both` df` and now `output_chunk` are referring to the same underlying table.

Comment (TomAugspurger) [reply python/cudf_polars/cudf_polars/streaming/actor_graph/repartition.py:231]: There's another one of these at https://github.com/rapidsai/cudf/pull/22651/changes#diff-032fc7e0d883bb28f32fe1e381474772938332ca9560f65c5c01834cbfd79322R229 too.

Also https://github.com/rapidsai/cudf/pull/22651/changes#diff-f1e7b7261ceb870ec382da613866d2b2c535403c1d2ca4f418661e637ba109c9R484.

**Review by TomAugspurger — APPROVED** [review]

File: python/cudf_polars/cudf_polars/streaming/actor_graph/repartition.py (hunk 5 above)
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/streaming/actor_graph/repartition.py:231]: TableChunks typically don't need to be deleted, because we relinquish ownership of the Table when we push it into the channel. I honestly don't know if it makes sense for us to have this `del df` here.

Comment (madsbk) [reply python/cudf_polars/cudf_polars/streaming/actor_graph/repartition.py:231]: Yes, we need to delete it to enable spilling ASAP.

First, we create a `TableChunk` that is a *view* on `df`. No data is copied, but the `TableChunk` keeps a reference to `df` to keep it alive.

```python
TableChunk.from_pylibcudf_table(
    df.table,
    df.stream,
    exclusive_view=True,
    br=context.br(),
)
```

By setting `exclusive_view=True`, we tell RapidsMPF that it may spill this `TableChunk`, and that doing so should free up memory. However, that is only true if the `TableChunk` is the sole remaining owner of `df`. 

Therefore, `del df` is needed to ensure that spilling the `TableChunk` actually frees up memory ASAP.

**Review by madsbk — APPROVED** [review]

File: python/cudf_polars/cudf_polars/streaming/actor_graph/repartition.py (hunk 5 above)
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/streaming/actor_graph/repartition.py:231]: Thanks for clarifying @madsbk ! That makes sense.
